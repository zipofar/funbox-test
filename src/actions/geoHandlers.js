import * as _ from "lodash";

const ymaps = window.ymaps;

export const getAddressByLetter = async (value) => {
    try {
        const result = await ymaps.suggest(value);
        return Promise.resolve(result);
    } catch (e) {
        return Promise.reject(e);
    }
};

export const getCoords = async (point) => {
    try {
        const pointAddress = point.value;
        const result = await ymaps.geocode(pointAddress, { results: 1 });
        const firstGeoObject = result.geoObjects.get(0);
        const coords = firstGeoObject.geometry.getCoordinates();

        const data = {
            displayName: point.displayName,
            value: pointAddress,
            coords: coords,
            id: _.uniqueId(),
        };
        return Promise.resolve(data);
    } catch (e) {
        return Promise.reject(e);
    }
};

export const getAddress = async (coords, idPoint, marker) => {
    try {
        const res = await ymaps.geocode(coords);
        const firstGeoObject = res.geoObjects.get(0);
        const data = {
            displayName: firstGeoObject.getAddressLine(),
            value: firstGeoObject.getCountry() + ', ' + firstGeoObject.getAddressLine(),
            coords: coords,
            id: idPoint,
        };

        marker.properties.set({
            balloonContent: firstGeoObject.getAddressLine(),
        });
        return Promise.resolve(data);
    } catch (e) {
        return Promise.reject(e);
    }
};