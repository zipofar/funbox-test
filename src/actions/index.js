import * as _ from 'lodash';

const ymaps = window.ymaps;

export const searchPoints = (value, cb) => async (dispatch) => {
    try {
        const result = await ymaps.suggest(value);
        cb(result);
    } catch (e) {

    }
};

export const removePointFromStore = (id) => ({ type: 'REMOVE_POINT', payload: id });
export const removePointDone = () => ({ type: 'REMOVE_POINT_DONE' });

export const reorderPointsDone = () => ({ type: 'REORDER_POINTS_DONE' });
export const reorderPointsInStore = (points) => ({ type: 'REORDER_POINTS', payload: points });

const addPointRequest = () => ({ type: 'ADD_POINT_REQUEST' });
const addPointSuccess = () => ({ type: 'ADD_POINT_SUCCESS' });
const addPointFailure = () => ({ type: 'ADD_POINT_FAILURE' });
export const addPointClear = () => ({ type: 'ADD_POINT_CLEAR' });
const addPointToStore = (data) => ({ type: 'ADD_POINT', payload: data });

export const addPoint = point => async (dispatch) => {
    dispatch(addPointRequest());
    try {
        const pointAddress = point.value;
        const result = await ymaps.geocode(pointAddress, { results: 1 });
        const firstGeoObject = result.geoObjects.get(0);
        const coords = firstGeoObject.geometry.getCoordinates();

        const data = {
            displayName: point.displayName,
            value: pointAddress,
            coords: coords,
            id: _.uniqueId(),
        };
        dispatch(addPointToStore(data));
        dispatch(addPointSuccess());
    } catch (e) {
        dispatch(addPointFailure());
    }
};

const updatePointRequest = () => ({ type: 'UPDATE_POINT_REQUEST' });
const updatePointSuccess = () => ({ type: 'UPDATE_POINT_SUCCESS' });
const updatePointFailure = () => ({ type: 'UPDATE_POINT_FAILURE' });
export const updatePointDone = () => ({ type: 'UPDATE_POINT_DONE' });
const updatePointToStore = (data) => ({ type: 'UPDATE_POINT', payload: data });

export const updatePoint = (coords, idPoint, marker) => async (dispatch) => {
    dispatch(updatePointRequest());
    try {
        const res = await ymaps.geocode(coords);
        const firstGeoObject = res.geoObjects.get(0);
        const data = {
            displayName: firstGeoObject.getAddressLine(),
            value: firstGeoObject.getCountry() + ', ' + firstGeoObject.getAddressLine(),
            coords: coords,
            id: idPoint,
        };

        marker.properties.set({
            balloonContent: firstGeoObject.getAddressLine(),
        });

        dispatch(updatePointToStore(data));
        dispatch(updatePointSuccess());
    } catch (e) {
        dispatch(updatePointFailure());
    }

};