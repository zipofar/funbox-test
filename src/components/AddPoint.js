import React from 'react';

export default class AddPoint extends React.Component
{
    state = {
        inputValue: '',
        searchedPoints: [],
        currentPoint: [],
        currentIndexSearch: null,
        hoverIndex: null,
    };

    onChangeValue = (e) => {
        const { value } = e.target;
        this.setState({ inputValue: value, currentIndexSearch: null });
        this.props.searchPoints(value, (points) => {this.setState({ searchedPoints: points })});
    };

    resetState = () => {
        this.setState({ inputValue: '', searchedPoints: [],  currentPoint: [] });
    };

    savePoint = () => {
        this.props.addPoint(this.state.currentPoint);
        this.resetState();
    };

    onSubmitPoint = (e) => {
        e.preventDefault();
        this.savePoint();
    };

    showDropDownSearch = () => {
        const { searchedPoints } = this.state;
        if (searchedPoints.length === 0) {
            return null;
        }
        return(
            <div className='col-sm-12'>
                <div id="searchedPoints" className='form-row'>
                    <div className="list-group">
                        { searchedPoints.map((item, i) => {
                            return(
                                <button
                                    key={i}
                                    type='button'
                                    className={`list-group-item ${ this.state.hoverIndex === i ? 'active' : '' }`}
                                    onClick={this.onClickDropDownPoints(item.displayName, i)}
                                    onMouseEnter={this.toggleHoverFindedItems(i)}
                                    onMouseLeave={this.toggleHoverFindedItems(i)}
                                >
                                    {item.displayName}
                                </button>
                            );
                        }) }
                    </div>
                </div>
            </div>
        );
    };

    toggleHoverFindedItems = index => (e) => {
        if (index === this.state.hoverIndex) {
            this.setState({ hoverIndex: null });
        } else {
            this.setState({
                hoverIndex: index,
                inputValue: this.state.searchedPoints[index].displayName,
                currentPoint: this.state.searchedPoints[index],
                currentIndexSearch: index,
            });
        }
    };

    onClickDropDownPoints = (value, index) => (e) => {
        e.preventDefault();
        this.setState({
            inputValue: value,
            currentPoint: this.state.searchedPoints[index],
            searchedPoints: [],
        }, this.savePoint);
    };

    onBlurPointInput = () => {
        setTimeout(() => this.resetState(), 400);
    };

    keyUpInputPoint = (e) => {
        if (e.key === 'Escape') {
            this.resetState();
        }
        if (e.key !== 'ArrowDown' && e.key !== 'ArrowUp') {
            return;
        }
        if (this.state.searchedPoints.length === 0) {
            return;
        }

        let currentIndex = this.state.currentIndexSearch;

        if (e.key === 'ArrowDown') {
            if (currentIndex === null) {
                currentIndex = 0;
            } else if (this.state.searchedPoints.length - 1 > currentIndex) {
                currentIndex += 1;
            }
        }

        if (e.key === 'ArrowUp') {
            if (currentIndex === null || currentIndex === 0) {
                return;
            } else if (currentIndex > 0) {
                currentIndex -= 1;
            }
        }

        this.setState({
            inputValue: this.state.searchedPoints[currentIndex].displayName,
            currentPoint: this.state.searchedPoints[currentIndex],
            hoverIndex: currentIndex,
            currentIndexSearch: currentIndex,
        });
    };

    render() {
        return(
            <div className='row'>
                <form onSubmit={this.onSubmitPoint} style={{width: '100%'}} autoComplete='off'>
                    <div className='form-row'>
                        <div className='col-sm-12'>
                            <input
                                type="text"
                                className="form-control"
                                id="inputPoint"
                                placeholder="New point route"
                                onChange={this.onChangeValue}
                                onBlur={this.onBlurPointInput}
                                onKeyUp={this.keyUpInputPoint}
                                value={this.state.inputValue}
                            />
                        </div>
                        {this.showDropDownSearch()}
                    </div>
                </form>
            </div>
        );
    }
}