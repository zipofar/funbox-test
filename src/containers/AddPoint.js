import { connect } from 'react-redux';
import AddPoint from '../components/AddPoint';
import * as actions from '../actions';

const mapStateToProps = state => {
    return {

    };
};

const container = connect(mapStateToProps, actions)(AddPoint);

export default container;