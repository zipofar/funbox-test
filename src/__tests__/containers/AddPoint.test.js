import React from 'react';
import { shallow, configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallowToJson } from 'enzyme-to-json';

import AddPoint from '../../containers/AddPoint';

import { Provider } from 'react-redux';
import { createStore } from "redux";
import RootReducer from "../../reducers";
import * as actions from '../../actions';

configure({ adapter: new Adapter() });

it('Should render AddPoint', () => {
    const store = createStore(RootReducer);
    mount(<Provider store = { store }><AddPoint /></Provider>);
});

it('Should match to snapshot AddPoint', () => {
    const store = createStore(RootReducer);
    const wrapper = mount(<Provider store = { store }><AddPoint /></Provider>);
    const divContainer = wrapper.find('#container-input-point');
    expect(shallowToJson(divContainer)).toMatchSnapshot();
});

describe('Should add two points to store', () => {

    const store = createStore(RootReducer);
    const coords1 = [55.76, 37.64];
    const coords2 = [10, 10];
    const wrapper = mount(<Provider store = { store }><AddPoint /></Provider>);
    const input = wrapper.find('#inputPoint');
    const form = wrapper.find('form');

    it('Should add first point to store', () => {
        const expectedPoints = [ { coords: coords1, namePoint: 'A', id: '1' } ];
        store.dispatch(actions.setCenterCoords(coords1));
        input.prop('onChange')({ target: { value: 'A' } });
        form.simulate('submit');

        expect(store.getState().points).toEqual(expectedPoints);
        expect(store.getState().addPointState).toEqual('success');
    });

    it('Should add second point to store', () => {
        const expectedPoints = [
            { coords: coords1, namePoint: 'A', id: '1' },
            { coords: coords2, namePoint: 'B', id: '2' },
        ];
        store.dispatch(actions.setCenterCoords(coords2));
        input.prop('onChange')({ target: { value: 'B' } });
        form.simulate('submit');

        expect(store.getState().points).toEqual(expectedPoints);
        expect(store.getState().addPointState).toEqual('success');
    });
});


/*
it('render AddPoint without crashing', () => {
    shallow(<AddPoint />);
});

it('write some symbol and call func "searchPoints"', () => {
    const comp = shallow(<AddPoint />);
    const input = comp.find('#inputPoint');
    input.prop('onChange')({ target: { value: 'A' } });
    expect(searchPoints).toBeCalled();
});
*/