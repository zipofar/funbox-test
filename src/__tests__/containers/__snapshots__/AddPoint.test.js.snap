// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match to snapshot AddPoint 1`] = `
<div
  className="row"
  id="container-input-point"
>
  <form
    autoComplete="off"
    onSubmit={[Function]}
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      className="form-row"
    >
      <div
        className="col-sm-12"
      >
        <input
          className="form-control"
          id="inputPoint"
          onChange={[Function]}
          placeholder="Enter name point"
          type="text"
          value=""
        />
      </div>
    </div>
  </form>
</div>
`;
